CONTRIBUTING.md

```diff
@@ -87,113 +85,154 @@ Usage: make <target>
+Markdown versions can be viewed in the `docs/source/markdown` directory.
+Files suffixed with `.in` are preliminary versions that are compiled into the final markdown files.
+
+## Libraries
+
+Podman uses a large amount of vendored library code, contained in the `vendor/` directory.
+This code is included in the Podman repository, but is actually maintained elsewhere.
+Pull requests that change the vendor/ directory directly will not be accepted.
+Instead, changes should be submitted to the original package (defined by the path in `vendor/`; for example, `vendor/github.com/containers/storage` is the [containers/storage library](https://github.com/containers/storage/).
+Once the changes have been merged into the original package, Podman's vendor directory can be updated by using `go get` on the appropriate version of the package, then running `make vendor-in-container`.
+
+## Testing
+
+Podman provides an extensive suite of regression tests in the `test/` directory.
+There is a [readme](https://github.com/containers/podman/blob/main/test/README.md) file available with details about the tests and how to run them.
+All pull requests should be accompanied by test changes covering the changes in the PR.
+Pull requests without tests cannot be merged without maintainer intervention.
+Maintainers will decide if tests are not necessary during review.
+
+### Types of Tests
+
+There are several types of tests run by Podman's upstream CI.
+* Build testing (including cross-build tests, and testing to verify each commit in a PR builds on its own)
+* Go format/lint checking
+* Unit testing
+* Integration testing (run on several operating systems, both root and rootless)
+* System testing (again, run on several operating systems, root and rootless)
+* API testing (validates the Podman REST API)
+* Machine testing (validates `podman machine` on Windows and Mac hosts)
+
+Changes will usually only need to be tested in one of these.
+For example, if you were to make a change to `podman run`, you could test this in either the system tests or the integration tests.
+It is not necessary to test a single change in multiple places.
+
+### Go Format and lint
+
+All code changes must pass ``make validatepr``.
+
+### Integration Tests
+
+Our primary means of performing integration testing for Podman is with the
+[Ginkgo](https://github.com/onsi/ginkgo) BDD testing framework. This allows
+us to use native Golang to perform our tests and there is a strong affiliation
+between Ginkgo and the Go test framework.  Adequate test cases are expected to
+be provided with PRs.
+
+For details on how to run the tests for Podman in your test environment, see the
+testing [README.md](test/README.md).
+
+### System Tests
+
+The system tests are written in Bash using the BATS framework.
+They provide less comprehensive coverage than the integration tests.
+They are intended to validate Podman builds before they are shipped by distributions.
```

Comment:
IF you document this here you properly want to show the directory path were they are defined.